workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ComposedViewFactory.kt

```diff
@@ -0,0 +1,141 @@
+import androidx.compose.ui.platform.ComposeView
+import com.squareup.workflow1.ui.LayoutRunner
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.ViewFactory
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.bindShowRendering
+import kotlin.reflect.KClass
+
+/**
+ * Creates a [ViewFactory] that uses a [Composable] function to display the rendering.
+ *
+ * Simple usage:
+ *
+ * ```
+ * val FooViewFactory = composedViewFactory { rendering, _ ->
+ *   Text(rendering.message)
+ * }
+ *
+ * …
+ *
+ * val viewRegistry = ViewRegistry(FooViewFactory, …)
+ * ```
+ *
+ * If you need to write a class instead of a function, for example to support dependency injection,
+ * see [ComposedViewFactory].
+ *
+ * For more details about how to write composable view factories, see [ComposedViewFactory].
+ */
+@WorkflowUiExperimentalApi
+public inline fun <reified RenderingT : Any> composedViewFactory(
+  noinline content: @Composable (
+    rendering: RenderingT,
+    environment: ViewEnvironment
+  ) -> Unit
+): ViewFactory<RenderingT> = composedViewFactory(RenderingT::class, content)
+
+@PublishedApi
+@WorkflowUiExperimentalApi
+internal fun <RenderingT : Any> composedViewFactory(
+  type: KClass<RenderingT>,
+  content: @Composable (
+    rendering: RenderingT,
+    environment: ViewEnvironment
+  ) -> Unit
+): ViewFactory<RenderingT> = object : ComposedViewFactory<RenderingT>() {
+  override val type: KClass<in RenderingT> = type
+  @Composable override fun Content(
+    rendering: RenderingT,
+    viewEnvironment: ViewEnvironment
+  ) {
+    content(rendering, viewEnvironment)
+  }
+}
+
+/**
+ * A [ViewFactory] that uses a [Composable] function to display the rendering. It is the composable
+ * version of [LayoutRunner].
+ *
+ * Simple usage:
+ *
+ * ```
+ * class FooViewFactory : ComposedViewFactory<Foo>() {
+ *   override val type = Foo::class
+ *
+ *   @Composable override fun Content(
+ *     rendering: Foo,
+ *     viewEnvironment: ViewEnvironment
+ *   ) {
+ *     Text(rendering.message)
+ *   }
+ * }
+ *
+ * …
+ *
+ * val viewRegistry = ViewRegistry(FooBinding, …)
+ * ```
+ *
+ * ## Nesting child renderings
+ *
+ * Workflows can render other workflows, and renderings from one workflow can contain renderings
+ * from other workflows. These renderings may all be bound to their own [ViewFactory]s. Regular
+ * [ViewFactory]s and `LayoutRunner`s use
+ * [WorkflowViewStub][com.squareup.workflow1.ui.WorkflowViewStub] to recursively show nested
+ * renderings using the [ViewRegistry][com.squareup.workflow1.ui.ViewRegistry].
+ *
+ * View factories defined using this function may also show nested renderings. Doing so is as simple
+ * as calling [WorkflowRendering] and passing in the nested rendering. See the kdoc on that function
+ * for an example.
+ *
+ * Nested renderings will have access to any ambients defined in outer composable, even if there are
```

Comment:
nit: Do Compose docs still use the term 'ambients' or has that been replaced by composition locals? (Disregard if you want to cast a wider net than composition locals, of course.)

Reply:
crap, thought i got these